####-sparse-observable.md

```diff
@@ -0,0 +1,517 @@
+This happens because the alphabet is formed of Hermitian operators (even though the complete observable need not be), but matrix multiplication does not preserve hermicity.
+
+This can in theory be avoided by storing an explicit $2\times2$ matrix of coefficients for single-qubit term.
+For observables whose terms all involve only one or two qubits, the memory impacts are potentially acceptable, but the constant-factor scaling as the number of non-identity qubits increases is far poorer overall.
+Asymptotically, doing this would have a memory usage of 36 bytes per non-identity qubit + 8 bytes per term, compared to 5 bytes per non-identity qubit + 16 bytes per term, and would correspondingly be more expensive to compute with and to decompose to "measurement" operations.
+
+Supporting things like this efficiently in the observable is also perhaps misleading for the user; the _measurement terms_ needed to support things like $\lvert 0\rangle\langle 1\rvert$ still involve the exponential explosion (they go like $X + iY$), which could easily lead to a user getting needing / being charged for exponentially scaling runtime from an operator that appeared benign.
+
+
+## Questions
+
+### Should `SparseObservable` allow parametric observables?
+
+The definition given of `SparseObservables` limits the coefficients to be double-precision complex floats.
+`SparsePauliOp` technically allows observables to be parametrised in terms of Qiskit's symbolic `Parameter` class (or really any Python object).
+
+Should `SparseObservable` also support these uses?
+
+Pros of support:
+
+* More flexibility in what can be represented in the operators.
+
+Cons:
+
+* Operations on the type will be more complex, and the general case will need to hold the Python GIL in Rust space to act on the observables.
+* More complex code is needed from Rust space to handle mathematical manipulations efficiently in the happy path of "no Python-space parametrisation".
+
+Jake: personally I'd avoid this unless we have a really strong compelling use-case for giving it first-class support.
+A user can always work around this simply by splitting the terms of their sum into different broadcast axes in the estimator pub, then calculating the sums themselves, which gives them far more freedom.
+
+
+### What mathematical operations should `SparseObservable` support?
+
+From above, the minimum set is:
+
+* addition of two `SparseObservable`s
+* tensor product of two `SparseObservable`s
+* evolution of one `SparseObservable` by a single term
+* multiplication by complex scalars
+* structural equality of two `SparseObservable`s (structural not mathematical; it's highly inefficient to detect equality if the abstract terms form an overcomplete spanning set).
+
+There are several others that might be possible too.
+These operations could all fairly easily be supported:
+
+* Evolution of `SparseObservable` by another: this is completely doable, just naturally has quadratic complexity.
```

Comment:
It comes up sometimes to compute the expectation value of the Hamiltonian squared in algos & applications (even in the estimator for the variance maybe?), so the would seem like an important feature

Reply:
Adding the expectation value computation for a given state is certainly easy enough, the trick is probably just finding some sensible representation of the statevector; we don't have the concept of a "sparse statevector" in Qiskit at the moment, and this operator is intended for numbers of qubits that dense statevectors can't represent.

Reply:
Hmm but if we square a Hamiltonian we might have to measure in new bases, e.g. if `H = X -> H^2 = I` we'd have to measure in Z-basis for the expval of `H^2` (ok maybe a bit basic this example... 😄). Or maybe I didn't correctly understand your proposal 🤔

Reply:
Oh sorry, I missed the word "squared" in your answer.

If we have some representation of a state, I think the expectation value of the square of a Hamiltonian might shake out neater in an API if we cast that problem to "evolve the (sparse) state by the operator, then take the inner product with itself"?

Reply:
Hm I'm not sure you can do that since you'd actually have to measure in different bases.. but we can also discuss this later 🙂